pkg/health/server/prober.go

```diff
@@ -234,34 +242,105 @@ func (p *prober) setNodes(added nodeMap, removed nodeMap) {
 	}
 }
 
-func (p *prober) updateIcmpStatus() {
-	p.Lock()
-	defer p.Unlock()
+const httpPathDescription = "Via L3"
 
-	for _, status := range p.results {
-		if status.Icmp == nil {
-			status.Icmp = &models.ConnectivityStatus{}
-			status.Icmp.Status = "Connection timed out"
+func (p *prober) getIPsByNode() map[string][]*net.IPAddr {
+	p.RLock()
+	defer p.RUnlock()
+
+	// p.nodes is mapped from all known IPs -> nodes in N:M configuration,
+	// so multiple IPs could refer to the same node. To ensure we only
+	// ping each node once, deduplicate nodes into map of nodeName -> []IP.
+	nodes := make(map[string][]*net.IPAddr)
+	for _, node := range p.nodes {
+		if nodes[node.Name] != nil {
+			// Already handled this node.
+			continue
+		}
+		nodes[node.Name] = []*net.IPAddr{}
+		for elem, primary := range node.Addresses() {
+			if _, addr := resolveIP(&node, elem, primary); addr != nil {
+				nodes[node.Name] = append(nodes[node.Name], addr)
+			}
 		}
 	}
+
+	return nodes
 }
 
-func (p *prober) clearIcmpStatus() {
-	p.Lock()
-	defer p.Unlock()
+func (p *prober) icmpPing(node string, ip string, ctx context.Context, resChan chan<- connectivityResult, wg *sync.WaitGroup) {
+	defer wg.Done()
+
+	result := &models.ConnectivityStatus{}
+	if err := p.icmpPingerRateLimiter.Wait(ctx); err != nil {
+		result.Status = err.Error()
+		resChan <- connectivityResult{ip: ip, status: result}
+		return
+	}
+
+	// Only add fields to the scoped logger if debug is enabled, to save on resources.
+	// This can be done since all logs in this function are debug-level only.
+	scopedLog := log
+	debugLogsEnabled := logging.CanLogAt(scopedLog.Logger, logrus.DebugLevel)
+	if debugLogsEnabled {
+		scopedLog = log.WithFields(logrus.Fields{
+			logfields.NodeName: node,
+			logfields.IPAddr:   ip,
+		})
+	}
+
+	scopedLog.Debug("Pinging host")
+	pinger, err := probing.NewPinger(ip)
+	if err != nil {
+		if debugLogsEnabled {
+			scopedLog.WithError(err).Debug("Failed to create pinger")
+		}
+		result.Status = err.Error()
+		resChan <- connectivityResult{ip: ip, status: result}
+		return
+	}
+
+	pinger.Timeout = p.server.ProbeDeadline
+	pinger.Count = 1
```

Comment:
note to myself: I see that we removed:
```
// FIXME: Doubling the default payload size to 16 is a workaround for GH-18177
	prober.Size = 2 * fastping.TimeSliceLength
```
The new implementation uses 24 bytes pings by default, so it should be fine. 
I will double-check that experimentally though :)